### PR TITLE
Onboarding: Weight Tracking [golden]

### PR DESCRIPTION
**Pull request by abhiuee**
## Objective
Add weight tracking support to the Pet entity in the Spring PetClinic REST API to enable veterinarians to monitor pet weight over time.

## Requirements

REST API Updates: All existing Pet REST endpoints must include weight field in JSON:

- GET /api/pets/{petId} - return weight as decimal number in JSON response
- PUT /api/pets/{petId} - accept weight as decimal number in JSON request body
- GET /api/pets - return weight for all pets in response array
- POST /api/owners/{ownerId}/pets - accept weight when creating pets
- PUT /api/owners/{ownerId}/pets/{petId} - accept weight in updates

JSON Format: Weight field should be represented as:

- Field name: weight
- Type: number (decimal)
- Optional: can be null for pets without recorded weight
- Example: "weight": 15.75 or "weight": null

## Acceptance Criteria

- All existing Pet REST endpoints include weight field in their JSON request/response payloads
- Weight field accepts decimal numbers and null values through REST API
- Weight field persists correctly - value set via POST/PUT can be retrieved via GET

## Technical Design

- Implementation: Add weight field to Pet entity, update DTO/mapper, modify database schema
- Testing: Create end-to-end tests that verify weight field behavior through the REST API endpoints listed in requirements.
- Tests must use actual HTTP requests to validate JSON request/response handling.

### FAIL_TO_PASS: 
org.springframework.samples.petclinic.rest.controller.integration.PetWeightIntegrationTests,
org.springframework.samples.petclinic.rest.controller.integration.OwnerPetWeightIntegrationTests

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -389,6 +389,12 @@
                                 <!-- switch documentation provider to springdoc -->
                                 <documentationProvider>springdoc</documentationProvider>
                             </configOptions>
+                            <typeMappings>
```
Comment (Gowrishbyndoor) [inline pom.xml]: These changes really needed? without this change can't we use BigDecimal type?

File: src/test/java/org/springframework/samples/petclinic/rest/controller/OwnerRestControllerTests.java [hunk 2]
```diff
@@ -28,6 +28,7 @@
 import org.springframework.samples.petclinic.mapper.PetMapper;
 import org.springframework.samples.petclinic.mapper.VisitMapper;
 import org.springframework.samples.petclinic.model.Owner;
+import org.springframework.samples.petclinic.model.Pet;
```
Comment (Gowrishbyndoor) [inline src/test/java/org/springframework/samples/petclinic/rest/controller/OwnerRestControllerTests.java]: Please write pure integration tests without any mocking, means your integration tests file should not depend on the Models, we do this way because if we copy the test folder completely to the agent branch it should not contains any implementation code dependencies.

File: src/test/java/org/springframework/samples/petclinic/rest/controller/OwnerRestControllerTests.java [hunk 3]
```diff
@@ -492,4 +496,125 @@ void testUpdateOwnersPetPetNotFound() throws Exception {
...
+        newPet.setId(999);
+        newPet.setWeight(new BigDecimal("20.50"));
+        ObjectMapper mapper = new ObjectMapper();
+        mapper.registerModule(new JavaTimeModule());
+        mapper.setDateFormat(new SimpleDateFormat("yyyy-MM-dd"));
+        mapper.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
+        String newPetAsJSON = mapper.writeValueAsString(newPet);
+        Pet pet = petMapper.toPet(newPet);
+        pet.setWeight(new BigDecimal("20.50"));
+        given(this.clinicService.findOwnerById(1)).willReturn(ownerMapper.toOwner(owners.get(0)));
```
Comment (Gowrishbyndoor) [inline src/test/java/org/springframework/samples/petclinic/rest/controller/OwnerRestControllerTests.java]: as mentioned above we should not mock the services

File: src/test/java/org/springframework/samples/petclinic/rest/controller/OwnerRestControllerTests.java [hunk 4]
```diff
@@ -492,4 +496,125 @@ void testUpdateOwnersPetPetNotFound() throws Exception {
...
+        pet.setWeight(null);
+        given(this.clinicService.findOwnerById(1)).willReturn(ownerMapper.toOwner(owners.get(0)));
+        doAnswer(invocation -> {
+            Pet savedPet = invocation.getArgument(0);
+            savedPet.setId(999);
+            return null;
+        }).when(this.clinicService).savePet(any(Pet.class));
+        this.mockMvc.perform(post("/api/owners/1/pets")
+                .content(newPetAsJSON).accept(MediaType.APPLICATION_JSON_VALUE).contentType(MediaType.APPLICATION_JSON_VALUE))
+            .andExpect(status().isCreated());
```
Comment (Gowrishbyndoor) [inline src/test/java/org/springframework/samples/petclinic/rest/controller/OwnerRestControllerTests.java]: I would suggests to call the GET API to check the values created/updated is available in DB. This comment you have to check in all test cases

File: src/test/java/org/springframework/samples/petclinic/rest/controller/PetWeightIntegrationTests.java [hunk 5]
```diff
@@ -0,0 +1,570 @@
...
+ */
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@ActiveProfiles({"h2", "spring-data-jpa"})
+@TestPropertySource(properties = {
+    "spring.sql.init.schema-locations=classpath*:db/h2/schema.sql",
+    "spring.sql.init.data-locations=classpath*:db/h2/data.sql",
+    "petclinic.security.enable=false",
+    "server.servlet.context-path="
+})
+class PetWeightIntegrationTests {
```
Comment (Gowrishbyndoor) [inline src/test/java/org/springframework/samples/petclinic/rest/controller/PetWeightIntegrationTests.java]: Instead of merging different APIs into one test file, to prevent duplicate tests, can we write all possible test in those 2 existing test files, as we already have the respective test files.

Also make sure that all given API should have all possible tests. Example POST api should have status 200/201, 400, 500 is possible.

**Review by Gowrishbyndoor — CHANGES_REQUESTED** [review]
Please note that golden PR and agent PR should have only one commit and agent with golden test should have 2 commits.

Also all your commit contains task description and PASS_TO_PASS or FAIL_TO_PASS criteria, please refer the instruction for more details.

File: src/test/java/org/springframework/samples/petclinic/rest/controller/OwnerRestControllerTests.java (hunk 2 above)
Comment (abhiuee) [reply src/test/java/org/springframework/samples/petclinic/rest/controller/OwnerRestControllerTests.java]: Integration tests are in a separate file, this test file only had unit tests so not worth writing integration tests in same file

File: src/test/java/org/springframework/samples/petclinic/rest/controller/OwnerRestControllerTests.java (hunk 4 above)
Comment (abhiuee) [reply src/test/java/org/springframework/samples/petclinic/rest/controller/OwnerRestControllerTests.java]: please see integration tests.

File: pom.xml (hunk 1 above)
Comment (abhiuee) [reply pom.xml]: this is fixed, thanks for catching this

**Comment by abhiuee** [comment]
@Gowrishbyndoor  since I have more than one commit in each, shall I create new PRs with just one commit each?

**Comment by Gowrishbyndoor** [comment]
> @Gowrishbyndoor since I have more than one commit in each, shall I create new PRs with just one commit each?

Please squash/rebase into 1 in all of your PR.

File: src/test/java/org/springframework/samples/petclinic/rest/controller/OwnerRestControllerTests.java (hunk 2 above)
Comment (Gowrishbyndoor) [reply src/test/java/org/springframework/samples/petclinic/rest/controller/OwnerRestControllerTests.java]: If integration tests covers these scenario then please remove it from unit tests, as it duplicate the tests, we need the unit tests when integration tests cannot cover, that case we need a mocked unit tests.

Also please create separate integration test for different types of API. PET related APIs separate and Owners related API separate.

Comment (abhiuee) [reply src/test/java/org/springframework/samples/petclinic/rest/controller/OwnerRestControllerTests.java]: Removed them from unit tests and created two separate integration test files

**Comment by abhiuee** [comment]
@Gowrishbyndoor  I have addressed PR comments, there is only one commit, two new integration test files are created, all tests are covered, no unit tests are touched. Please review it, once approved I will fix the agent task with golden tests as well.

File: src/test/java/org/springframework/samples/petclinic/rest/controller/PetWeightIntegrationTests.java (hunk 5 above)
Comment (abhiuee) [reply src/test/java/org/springframework/samples/petclinic/rest/controller/PetWeightIntegrationTests.java]: This is done.

File: src/test/java/org/springframework/samples/petclinic/rest/controller/integration/OwnerPetWeightIntegrationTests.java [hunk 6]
```diff
@@ -0,0 +1,538 @@
...
+        assertEquals(new BigDecimal("15.75"), response.getBody().getWeight());
+    }
+
+    // ========== Error Scenario Tests (400, 404, 500) ==========
+
+    @Test
+    void testCreatePetWithMissingRequiredFields() {
+        // POST /api/owners/{ownerId}/pets - 400 when required fields are missing
+        PetFieldsDto petFields = new PetFieldsDto();
+        petFields.setName(null); // Missing required field
```
Comment (Gowrishbyndoor) [inline src/test/java/org/springframework/samples/petclinic/rest/controller/integration/OwnerPetWeightIntegrationTests.java]: This is different use case tests, for your requirement it won't make value addition to your tests.

For your requirement you should have tests only related to weight field. instead of this test you can a tests with negative weight that will cover the 400 scenario.

Please remove all unnecessary tests which is not part of your requirement.

File: src/test/java/org/springframework/samples/petclinic/rest/controller/integration/OwnerPetWeightIntegrationTests.java [hunk 7]
```diff
@@ -0,0 +1,538 @@
...
+
+        assertEquals(HttpStatus.OK, getResponse.getStatusCode());
+        assertNotNull(getResponse.getBody());
+        assertEquals(createdPetId, getResponse.getBody().getId());
+        assertEquals("Charlie", getResponse.getBody().getName());
+        assertEquals(new BigDecimal("22.75"), getResponse.getBody().getWeight());
+    }
+
+    @Test
+    void testWeightAsDecimalNumber() throws Exception {
```
Comment (Gowrishbyndoor) [inline src/test/java/org/springframework/samples/petclinic/rest/controller/integration/OwnerPetWeightIntegrationTests.java]: I would suggest not to create too many tests for the same scenarios, in this case you have already a test for success scenario, this tests similar to that so it won't add any values to the tests, just a duplicate tests.

File: src/test/java/org/springframework/samples/petclinic/rest/controller/integration/PetWeightIntegrationTests.java [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+            HttpMethod.PUT,
+            entity,
+            String.class
+        );
+
+        assertEquals(HttpStatus.NOT_FOUND, response.getStatusCode());
+    }
+
+    @Test
+    void testUpdatePetWithInvalidData() {
```
Comment (Gowrishbyndoor) [inline src/test/java/org/springframework/samples/petclinic/rest/controller/integration/PetWeightIntegrationTests.java]: Here also please remove all tests which is not related to your requirement.

File: src/test/java/org/springframework/samples/petclinic/rest/controller/integration/OwnerPetWeightIntegrationTests.java [hunk 9]
```diff
@@ -0,0 +1,538 @@
...
+package org.springframework.samples.petclinic.rest.controller.integration;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.web.client.TestRestTemplate;
+import org.springframework.boot.test.web.server.LocalServerPort;
+import org.springframework.http.*;
+import org.springframework.samples.petclinic.model.Owner;
```
Comment (Gowrishbyndoor) [inline src/test/java/org/springframework/samples/petclinic/rest/controller/integration/OwnerPetWeightIntegrationTests.java]: Please take a note always write integration tests without importing the model in tests file because when we copy the test folder to the agent branch that time it may fail because of models dependencies. 

In the on boarding requirement it might work but in the production task it might not work so.

Its a suggestion for now, no need to change this now. Please remember it for production tasks.

**Review by Gowrishbyndoor — CHANGES_REQUESTED** [review]
Only remove unnecessary test which is not part of your requirement other than that PR looks good to me.

File: src/test/java/org/springframework/samples/petclinic/rest/controller/PetWeightIntegrationTests.java (hunk 5 above)
Comment (Gowrishbyndoor) [reply src/test/java/org/springframework/samples/petclinic/rest/controller/PetWeightIntegrationTests.java]: > Also make sure that all given API should have all possible tests. Example POST api should have status 200/201, 400, 500 is possible.

Sorry my comment may be guided you to wrong direction, i mean task description mentioned api should have all possible tests which should cover only mentioned details in the tasks. in this case you should write tests cases for weight field only but all POST, PUT, GET api should be covered.

**Comment by abhiuee** [comment]
> Only remove unnecessary test which is not part of your requirement other than that PR looks good to me.

Removed all the tests that are not necessary and not part of the requirement

**Review by Gowrishbyndoor — APPROVED** [review]
